File: tests/helpers/fixtures.ts
import { mkdtemp } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { saveRegistry } from "../../src/registry/persist.ts";
import type { Registry } from "../../src/registry/types.ts";

// Two plaintext local vaults, two single-mode consumers, one group. Tests
// mutate copies via overrides or direct assignment on the returned object.
export function makeRegistry(overrides: Partial<Registry> = {}): Registry {
  return {
    schemaVersion: 2,
    defaults: { vault: "local" },
    vaults: {
      local: { vaultType: "menv-local", vaultConfig: { filename: ".menv/vault.json", encryption: false } },
      production: {
        vaultType: "menv-local",
        vaultConfig: { filename: ".menv/vault.production.json", encryption: false },
      },
    },
    consumers: {
      api: { strategyType: "single", strategyConfig: { baseDir: "apps/api", filename: ".env" } },
      web: { strategyType: "single", strategyConfig: { baseDir: "apps/web", filename: ".env" } },
    },
    groups: { db: { title: "Database" } },
    globals: {},
    variables: {},
    compose: { files: [] },
    ...overrides,
  };
}

export async function tmpRepo(registry?: Registry): Promise<string> {
  const root = await mkdtemp(join(tmpdir(), "menv-cli-"));
  if (registry) await saveRegistry(root, registry);
  return root;
}

File: go/tests/helpers/fixtures.go
package helpers

import (
	"encoding/json"
	"os"
	"testing"

	"github.com/nikrabaev/menv/internal/registry"
)

// MakeRegistry returns a baseline 2-vault, 2-consumer, 1-group registry.
// Tests mutate copies via field assignment or overrides.
func MakeRegistry() registry.Registry {
	localConfig, _ := json.Marshal(map[string]any{
		"filename":   ".menv/vault.json",
		"encryption": false,
	})
	prodConfig, _ := json.Marshal(map[string]any{
		"filename":   ".menv/vault.production.json",
		"encryption": false,
	})
	return registry.Registry{
		SchemaVersion: 2,
		Defaults:      registry.Defaults{Vault: "local"},
		Vaults: map[string]registry.VaultDef{
			"local": {
				VaultType:   "menv-local",
				VaultConfig: json.RawMessage(localConfig),
			},
			"production": {
				VaultType:   "menv-local",
				VaultConfig: json.RawMessage(prodConfig),
			},
		},
		Consumers: map[string]registry.ConsumerDef{
			"api": {
				StrategyType: "single",
				StrategyConfig: registry.StrategyConfig{
					BaseDir:  "apps/api",
					Filename: ".env",
				},
			},
			"web": {
				StrategyType: "single",
				StrategyConfig: registry.StrategyConfig{
					BaseDir:  "apps/web",
					Filename: ".env",
				},
			},
		},
		Groups: map[string]registry.GroupDef{
			"db": {Title: "Database"},
		},
		Globals:   map[string]registry.GlobalDef{},
		Variables: map[string]registry.VariableDef{},
		Compose:   registry.Compose{Files: []string{}},
	}
}

// TmpRepo creates a temporary directory, optionally saving a registry into it.
// Calls t.Cleanup to remove the directory when the test ends.
func TmpRepo(t *testing.T, r *registry.Registry) string {
	t.Helper()
	dir, err := os.MkdirTemp("", "menv-test-")
	if err != nil {
		t.Fatalf("TmpRepo: %v", err)
	}
	t.Cleanup(func() { os.RemoveAll(dir) })
	if r != nil {
		if err := registry.SaveRegistry(dir, *r); err != nil {
			t.Fatalf("TmpRepo SaveRegistry: %v", err)
		}
	}
	return dir
}
